parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,421 @@
+        examples = self.metrics['examples']
+        if examples > 0:
+            m['examples'] = examples
+            m['loss'] = self.metrics['loss']
+            m['mean_loss'] = self.metrics['loss'] / examples
+
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            base[k] = round_sigfigs(v, 4)
+        return base
+
+    def train_step(self, batch):
+
+        tensors = self._vectorize(batch['observations'])
+        loss = self.model(*tensors)
+        self.metrics['examples'] += len(batch['observations'])
+        self.metrics['loss'] += loss
+
+        self.backward(loss)
+        self.update_params()
+        #self.optimizer.step()
+        self.zero_grad()
+        #self.optimizer.zero_grad()
+
+        # # predictions
+        # with torch.no_grad():
+        #     b_input_ids, b_segment_ids, b_input_mask, _, _ = tensors
+        #     print(b_input_ids)
+        #     input(...)
+        #     b_start_logits, b_end_logits = self.model(b_input_ids,
+        #                                               b_segment_ids,
+        #                                               b_input_mask)
+
+        # predictions = []
+        # for start_logits, end_logits, input_ids in zip(b_start_logits,
+        #                                                b_end_logits,
+        #                                                b_input_ids):
+        #     prediction = self._get_prediction(start_logits, end_logits,
+        #                                       input_ids)
+        #     predictions.append(prediction)
+
+        # return Output(predictions)
+
+    def eval_step(self, batch):
+
+        segment_ids = torch.zeros(batch['text_vec'].size(),
+                                  dtype=torch.long).to(self.device)
+        input_mask = batch['text_vec'].clone()
+        input_mask[input_mask != self.dict.pad_idx] = 1
+
+        with torch.no_grad():
+            b_start_logits, b_end_logits = self.model(batch['text_vec'],
+                                                      segment_ids, input_mask)
+
+        predictions = []
+        for start_logits, end_logits, input_ids in zip(b_start_logits,
+                                                       b_end_logits,
+                                                       batch['text_vec']):
+            prediction = self._get_prediction(start_logits, end_logits,
+                                              input_ids)
+            predictions.append(prediction)
+
+        return Output(predictions)
+
+    def share(self):
+        """Share model parameters."""
+        shared = super().share()
+        shared['model'] = self.model
+        shared['metrics'] = self.metrics
+        return shared
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        TorchAgent.add_cmdline_args(parser)
+        parser = parser.add_argument_group('BERT Classifier Arguments')
+        parser.add_argument('--type-optimization',
```

Comment:
I wouldn't mind retiring this argument. We haven't seen an instance where you only want to fine tune a subset and it actually work?